File: go-backend-interview-prep/08_http_and_backend/04_request_parsing_and_validation/main.py
"""Request parsing and validation -- Python equivalent of the Go example."""

import json
import threading
import urllib.request
import urllib.error
from http.server import HTTPServer, BaseHTTPRequestHandler


def json_response(handler, status: int, data: dict) -> None:
    handler.send_response(status)
    handler.send_header("Content-Type", "application/json")
    handler.end_headers()
    handler.wfile.write(json.dumps(data).encode())


class Handler(BaseHTTPRequestHandler):
    def do_POST(self):
        if self.path == "/users":
            self._create_user()
            return
        json_response(self, 404, {"error": "not found"})

    def do_GET(self):
        if self.path.startswith("/search"):
            self._search()
            return
        json_response(self, 404, {"error": "not found"})

    def _create_user(self):
        # Check Content-Type
        ct = self.headers.get("Content-Type", "")
        if "application/json" not in ct:
            json_response(self, 400, {"error": "Content-Type must be application/json"})
            return

        # Decode JSON
        length = int(self.headers.get("Content-Length", 0))
        if length > 1_048_576:  # 1 MB limit
            json_response(self, 400, {"error": "body too large"})
            return

        try:
            body = json.loads(self.rfile.read(length))
        except json.JSONDecodeError as e:
            json_response(self, 400, {"error": f"invalid JSON: {e}"})
            return

        # Validate fields
        errors = []
        name = body.get("name", "")
        email = body.get("email", "")
        if not name:
            errors.append("name is required")
        if not email:
            errors.append("email is required")
        elif "@" not in email:
            errors.append("email must contain @")

        if errors:
            json_response(self, 400, {"error": "validation failed", "fields": errors})
            return

        json_response(self, 201, {"message": "user created", "name": name, "email": email})

    def _search(self):
        from urllib.parse import urlparse, parse_qs
        qs = parse_qs(urlparse(self.path).query)
        page = qs.get("page", ["1"])[0]
        limit = qs.get("limit", ["10"])[0]
        json_response(self, 200, {"page": page, "limit": limit})

    def log_message(self, *args):
        pass


def main() -> None:
    server = HTTPServer(("127.0.0.1", 0), Handler)
    host, port = server.server_address
    base = f"http://{host}:{port}"
    print(f"Server on {base}")
    threading.Thread(target=server.serve_forever, daemon=True).start()

    print("\n--- Demo ---")

    # Valid request
    req = urllib.request.Request(
        f"{base}/users",
        data=json.dumps({"name": "Alice", "email": "alice@example.com"}).encode(),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    resp = urllib.request.urlopen(req)
    print(f"  valid:     {resp.read().decode()}")

    # Missing fields
    req2 = urllib.request.Request(
        f"{base}/users",
        data=b"{}",
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        urllib.request.urlopen(req2)
    except urllib.error.HTTPError as e:
        print(f"  invalid:   {e.read().decode()}")

    # Bad email
    req3 = urllib.request.Request(
        f"{base}/users",
        data=json.dumps({"name": "Bob", "email": "nope"}).encode(),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        urllib.request.urlopen(req3)
    except urllib.error.HTTPError as e:
        print(f"  bad email: {e.read().decode()}")

    # Query params
    resp4 = urllib.request.urlopen(f"{base}/search?page=2&limit=25")
    print(f"  query:     {resp4.read().decode()}")

    server.shutdown()
    print("Done.")


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/08_http_and_backend/04_request_parsing_and_validation/main.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net"
	"net/http"
	"strings"
	"time"
)

type UserRequest struct {
	Name  string `json:"name"`
	Email string `json:"email"`
}

func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(v)
}

func createUserHandler(w http.ResponseWriter, r *http.Request) {
	// Limit body size (1 MB)
	r.Body = http.MaxBytesReader(w, r.Body, 1<<20)

	// Check Content-Type
	ct := r.Header.Get("Content-Type")
	if !strings.HasPrefix(ct, "application/json") {
		writeJSON(w, http.StatusBadRequest, map[string]string{
			"error": "Content-Type must be application/json",
		})
		return
	}

	// Decode JSON
	var req UserRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{
			"error": "invalid JSON: " + err.Error(),
		})
		return
	}

	// Validate fields
	errors := []string{}
	if req.Name == "" {
		errors = append(errors, "name is required")
	}
	if req.Email == "" {
		errors = append(errors, "email is required")
	} else if !strings.Contains(req.Email, "@") {
		errors = append(errors, "email must contain @")
	}
	if len(errors) > 0 {
		writeJSON(w, http.StatusBadRequest, map[string]any{
			"error":  "validation failed",
			"fields": errors,
		})
		return
	}

	// Success
	writeJSON(w, http.StatusCreated, map[string]string{
		"message": "user created",
		"name":    req.Name,
		"email":   req.Email,
	})
}

// queryHandler shows query param parsing
func queryHandler(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query()
	page := q.Get("page")
	limit := q.Get("limit")
	if page == "" {
		page = "1"
	}
	if limit == "" {
		limit = "10"
	}
	writeJSON(w, http.StatusOK, map[string]string{
		"page":  page,
		"limit": limit,
	})
}

func main() {
	mux := http.NewServeMux()
	mux.HandleFunc("POST /users", createUserHandler)
	mux.HandleFunc("GET /search", queryHandler)

	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		log.Fatal(err)
	}
	addr := ln.Addr().String()
	fmt.Printf("Server on http://%s\n", addr)
	fmt.Println("  POST /users   -> create user (JSON body)")
	fmt.Println("  GET  /search  -> query params demo")

	server := &http.Server{Handler: mux}

	// Self-test
	go func() {
		time.Sleep(100 * time.Millisecond)
		base := "http://" + addr
		fmt.Println("\n--- Demo ---")

		// Valid request
		body := `{"name":"Alice","email":"alice@example.com"}`
		resp, _ := http.Post(base+"/users", "application/json", strings.NewReader(body))
		if resp != nil {
			var result map[string]any
			json.NewDecoder(resp.Body).Decode(&result)
			resp.Body.Close()
			fmt.Printf("  valid:   %v\n", result)
		}

		// Missing fields
		resp2, _ := http.Post(base+"/users", "application/json", strings.NewReader(`{}`))
		if resp2 != nil {
			var result map[string]any
			json.NewDecoder(resp2.Body).Decode(&result)
			resp2.Body.Close()
			fmt.Printf("  invalid: %v\n", result)
		}

		// Bad email
		resp3, _ := http.Post(base+"/users", "application/json", strings.NewReader(`{"name":"Bob","email":"nope"}`))
		if resp3 != nil {
			var result map[string]any
			json.NewDecoder(resp3.Body).Decode(&result)
			resp3.Body.Close()
			fmt.Printf("  bad email: %v\n", result)
		}

		// Query params
		resp4, _ := http.Get(base + "/search?page=2&limit=25")
		if resp4 != nil {
			var result map[string]any
			json.NewDecoder(resp4.Body).Decode(&result)
			resp4.Body.Close()
			fmt.Printf("  query:   %v\n", result)
		}

		fmt.Println("\nDemo done.")
		server.Close()
	}()

	if err := server.Serve(ln); err != http.ErrServerClosed {
		log.Fatal(err)
	}
}
